pkg/controller/garbagecollector/propagator.go

```diff
@@ -0,0 +1,356 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollector
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/v1"
+	metav1 "k8s.io/kubernetes/pkg/apis/meta/v1"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+)
+
+type eventType int
+
+const (
+	addEvent eventType = iota
+	updateEvent
+	deleteEvent
+)
+
+type event struct {
+	eventType eventType
+	obj       interface{}
+	// the update event comes with an old object, but it's not used by the garbage collector.
+	oldObj interface{}
+}
+
+// Propagator: based on the events supplied by the informers, Propagator updates
+// uidToNode, a graph that caches the dependencies as we know, and enqueues
+// items to the dirtyQueue and orphanQueue.
+type Propagator struct {
+	eventQueue *workqueue.TimedWorkQueue
+	// uidToNode doesn't require a lock to protect, because only the
+	// single-threaded Propagator.processEvent() reads/writes it.
+	uidToNode *concurrentUIDToNode
+	gc        *GarbageCollector
```

Comment:
How about let GC pass the dirtyQueue and orphanQueue as arguments to the propagator, then we won't have the cycle?